### ISSUE TITLE
Get the span of a parser match?

### ISSUE THREAD

**Issue by jods4** [benjamin-hodgson/Pidgin#40]
It would be nice to be able to have as a result a `Span` in the original string/buffer that covers the matched extend.

## Use cases
1. Re-use some very efficient parsers once I know a substring matches the expected format.

It is quite a lot of code to assemble some literal types such as `double`, `DateTimeOffset` or `TimeSpan`, especially if you're going to support all optional parts and variations.

The framework already has some very efficient parsers for those and it's relatively easy to just check if the input matches a format you're supporting, then letting the framework do the actual parsing.

2. Sometimes you want to validate a complex format, but then just keep the result as a string.
Of course, you can Concat all the parts, but it seems to me it would be easier and more efficient to just skip intermediate results and get the span in the original string.

## Proposed API
`Predicate.SpanResult<T>(Func<Span<string>, T> selector)`

## Examples
Want to parse a phone number literal that should look `(0xx) xx xx`? and then keep it as a string?
```csharp
String("(0")
  .Then(Digit.Repeat(2))
  .Then(String(") ")
  .Then(Digit.Repeat(2))
  .Then(Char(' '))
  .Then(Digit.Repeat(2))
  .SpanResult(s => s.ToString()); // Result is the full matched string
```
Want to parse a decimal?
```csharp
Char('-').Optional()
  .Then(Digit.SkipAtLeastOnce())
  .Then(Char('.').Then(Digit.SkipMany()).Optional())
  .SpanResult(s => decimal.Parse(s)); // Result is the parsed decimal number
```
Compare that with the work required to build up the decimal yourself.

## New parser: Regex
Based on that idea that sometimes you just want to grab the matched substring, I think a regex parser building block that returns a `Span` when it matches would make the scenarios above even simpler.

```csharp
var phoneParser = Regex(@"\(0\d\d) \d\d \d\d");

var decimalParser = Regex(@"-?\d+(\.\d*)?", s => decimal.Parse(s));
```

**Comment by benjamin-hodgson** [comment]
I outlined in https://github.com/benjamin-hodgson/Pidgin/issues/8 why a `Span` API is not really feasible for a high-level parser combinator library. At the very least there would need to be a large-scale redesign of the whole API (and a corresponding v3.0 release). I'm certainly not against that in principle but there's a **lot** of details to work out before implementation.

**Comment by jods4** [comment]
I don't see how the argument in #8 applies here?

From #8:
> The problem is that there's no distinction between parsers which return data directly from the input stream (like Char and String) and parsers which return some calculated value.

What I'm suggesting is totally unrelated to return values of parsers. In fact all parsers could just be skipped.

Basically I'm suggesting that a Parser wrapped with `SpanResult` would:

1. Take note of current (starting) position in input stream.
2. Call parsers as usual, ignoring results.
3. If parsing was unsuccessful, abort. Otherwise continue.
4. Take note of current (ending) position in input stream.
5. Call `SpanResult` delegate with `Span(starting, ending)`. Note that the lifetime of Span is perfect, no problem here.
6. Final result is the value returned by 5.

I agree with you that exposing `ParserState` would be interesting. There's no reason I couldn't (try to) create those parser combinators myself, except `Parser` can't be derived outside of your assembly.

**Comment by benjamin-hodgson** [comment]
Ah, I see what you mean. I thought you were talking about a `Span` which contained the _results_ of a child parser, rather than the dynamic extent of the input which was consumed by a child parser.

As a method on `Parser` itself, that doesn't seem to fit the "least surprise principle"; all the other `Parser` methods operate on their argument's result (not its input). There's a bunch of functionality in `Parser` which would effectively be ignored by `SpanResult` (`Select` etc), which seems counterintuitive given how the library works in general.

I'd probably design this with a separate `Pattern` type, with combinators for defining how to match a sequence of tokens from the input but _no methods for computing intermediate results_.

```csharp
public delegate T SpanDelegate<TToken, T>(ReadOnlySpan<TToken> extent);
public abstract class Pattern<TToken>
{
    public static Pattern<TToken> Token(Func<TToken, bool> match);

    public static Pattern<TToken> Empty { get; }
    public static Pattern<TToken> Fail { get; }
    public Pattern<TToken> Then(Pattern<TToken> next);
    public Pattern<TToken> Or(Pattern<TToken> next);
    // etc. But no Select.

    public Parser<TToken, T> SpanResult(SpanDelegate<TToken, T> resultSelector);
}
```

So it'd be a monoidal API, rather than a monadic one. One missing part of this API is `Then(Func<T, Parser<TToken, U>> next)` - you can't dynamically compute the next piece of the pattern given the result of the first piece.

I suppose `Pattern` is ultimately a combinator-oriented way to write a regular expression. So perhaps it should just be

```
public static partial class Parser<TToken>
{
    public static Parser<TToken, T> Regex<T>(string pattern, SpanDelegate<TToken, T> resultSelector);
}
```

, where `pattern` is some suitable subset of the usual regular expression syntax. What do you think?

**Comment by benjamin-hodgson** [comment]
This would make #23 cleaner and more efficient.

**Comment by jods4** [comment]
Having a distinct API set that doesn't care about intermediate results would be even more efficient but it also means duplicating all the existing building blocks? 😐

**Comment by benjamin-hodgson** [comment]
I don't mind _implementing_ extra stuff to support `Pattern`. Do you think the API is confusing? Client code would _look_ the same as `Parser` code (assuming the method names match), so many of the programming concepts would be transferrable. Perhaps it belongs in a separate namespace?

**Comment by benjamin-hodgson** [comment]
It'd potentially break certain `using static` use cases.

**Comment by benjamin-hodgson** [comment]
I've put together an initial implementation of this idea in #41. See what you think.

**Comment by benjamin-hodgson** [comment]
Merged #46. Thanks for the suggestion!

**Comment by benjamin-hodgson** [comment]
Shipped in [v2.2.0](https://www.nuget.org/packages/Pidgin/2.2.0)
